node-metrics/src/api/node_validator/v0/create_node_validator_api.rs

```diff
@@ -0,0 +1,472 @@
+        K1: Sink<Url, Error = SendError> + Send + Unpin + 'static,
+        K2: Sink<Leaf<SeqTypes>, Error = SendError> + Send + Unpin + 'static,
+    {
+        let task_handle = async_std::task::spawn(Self::process_messages(
+            event_stream,
+            url_sender,
+            leaf_sender,
+        ));
+
+        Self {
+            task_handle: Some(task_handle),
+        }
+    }
+
+    /// [process_messages] is a function that will process messages from the
+    /// provided event stream.
+    async fn process_messages<S, K1, K2>(event_receiver: S, url_sender: K1, leaf_sender: K2)
+    where
+        S: Stream<Item = Event<SeqTypes>> + Send + Unpin + 'static,
+        K1: Sink<Url, Error = SendError> + Unpin,
+        K2: Sink<Leaf<SeqTypes>, Error = SendError> + Unpin,
+    {
+        let mut event_stream = event_receiver;
+        let mut url_sender = url_sender;
+        let mut leaf_sender = leaf_sender;
+        loop {
+            let event_result = event_stream.next().await;
+            let event = match event_result {
+                Some(event) => event,
+                None => {
+                    tracing::info!("event stream closed");
+                    break;
+                }
+            };
+
+            let Event { event, .. } = event;
+
+            match event {
+                EventType::Decide { leaf_chain, .. } => {
+                    for leaf_info in leaf_chain.iter() {
+                        let leaf = leaf_info.leaf.clone();
+
+                        let send_result = leaf_sender.send(leaf).await;
+                        if let Err(err) = send_result {
+                            tracing::info!("leaf sender closed: {}", err);
+                            return;
+                        }
+                    }
+                }
+
+                EventType::ExternalMessageReceived(external_message_bytes) => {
+                    let roll_call_info = match bincode::deserialize(&external_message_bytes) {
+                        Ok(ExternalMessage::RollCallResponse(roll_call_info)) => roll_call_info,
+
+                        Err(err) => {
+                            tracing::info!(
+                                "failed to deserialize external message, unrecognized: {}",
+                                err
+                            );
+                            continue;
+                        }
+
+                        _ => {
+                            // Ignore any other potentially recognized messages
+                            continue;
+                        }
+                    };
+
+                    let public_api_url = roll_call_info.public_api_url;
+
+                    // Send the the discovered public url to the sink
+                    let send_result = url_sender.send(public_api_url).await;
+                    if let Err(err) = send_result {
+                        tracing::info!("url sender closed: {}", err);
+                        return;
```

Comment:
I think this could make more sense than the previous one.  I think `Leaf`s are more important than the `Url`s in this case.

Reply:
Addressed in commit ea3d37a4fed07ab35566b3e651f4bd4c664fc813.

Reply:
Tbh after reading your previous comment I would put this back to returning/panicking rather than trying to continue in a partially broken state. Even if leaves are more important, URLs are still important....I'm just thinking about how this would manifest and it seems like we would just see the dashboard stop updating node identity information silently, so it might not even be obvious that anything is broken, we'd just be looking at stale data. I definitely want to avoid that

Reply:
Resolved in c0206d153d4d7ed2bbf676954ca6758123e24085.